pkg/restapi/comparator/operation/authz.go

```diff
@@ -0,0 +1,199 @@
+
+		return
+	}
+
+	edvURL, err := url.Parse(docMeta.URI)
+	if err != nil {
+		respondErrorf(w, http.StatusInternalServerError, "failed to parse doc uri: %s", err.Error())
+
+		return
+	}
+
+	parts := strings.Split(docMeta.URI, "/")
+
+	vaultID := parts[len(parts)-3]
+	docID := parts[len(parts)-1]
+
+	response, err := o.cshClient.PostHubstoreProfilesProfileIDQueries(
+		operations.NewPostHubstoreProfilesProfileIDQueriesParams().
+			WithTimeout(requestTimeout).
+			WithProfileID(o.cshProfile.ID).
+			WithRequest(&cshclientmodels.DocQuery{VaultID: &vaultID, DocID: &docID,
+				UpstreamAuth: &cshclientmodels.DocQueryAO1UpstreamAuth{
+					Edv: &cshclientmodels.UpstreamAuthorization{
+						BaseURL: fmt.Sprintf("%s://%s/%s", edvURL.Scheme, edvURL.Host, parts[3]),
+						Zcap:    authz.Scope.AuthTokens.Edv,
+					},
+					Kms: &cshclientmodels.UpstreamAuthorization{
+						BaseURL: fmt.Sprintf("%s://%s", kmsURL.Scheme, kmsURL.Host),
+						Zcap:    authz.Scope.AuthTokens.Kms,
+					},
+				}}))
+	if err != nil {
+		respondErrorf(w, http.StatusInternalServerError, "failed to create query: %s", err.Error())
+
+		return
+	}
+
+	// deriving a child zcap for csh
+	zcap, err := o.driveZCAPForCSH(*authz.RequestingParty, response.Location,
+		authz.Scope.Caveats())
+	if err != nil {
+		respondErrorf(w, http.StatusInternalServerError, "failed to drive child zcap from csh zcap: %s", err.Error())
+
+		return
+	}
+
+	authToken, err := gzipThenBase64URL(zcap)
+	if err != nil {
+		respondErrorf(w, http.StatusInternalServerError, "failed to compress zcap: %s", err.Error())
+
+		return
+	}
+
+	headers := map[string]string{
+		"Content-Type": "application/json",
+	}
+
+	respond(w, http.StatusOK, headers, models.Authorization{RequestingParty: authz.RequestingParty,
+		AuthToken: authToken})
+}
+
+func (o *Operation) driveZCAPForCSH(invokerDID, queryIDPath string,
+	caveats []models.Caveat) (*zcapld.Capability, error) {
+	cshZCAP, err := parseCompressedZCAP(o.cshProfile.Zcap)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse CHS profile zcap: %w", err)
+	}
+
+	keyID, key, err := getKey(o.comparatorConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return zcapld.NewCapability(&zcapld.Signer{
+		SignatureSuite:     ed25519signature2018.New(suite.WithSigner(&ed25519Signer{key: key})),
+		SuiteType:          ed25519signature2018.SignatureType,
+		VerificationMethod: fmt.Sprintf("%s#%s", *o.comparatorConfig.Did, keyID),
+		ProcessorOpts:      []jsonld.ProcessorOpts{jsonld.WithDocumentLoader(verifiable.CachingJSONLDLoader())},
+	}, zcapld.WithParent(cshZCAP.ID), zcapld.WithInvoker(invokerDID),
+		zcapld.WithAllowedActions("reference"),
+		zcapld.WithCaveats(toZCaveats(caveats)...),
+		zcapld.WithInvocationTarget(queryIDPath, "urn:confidentialstoragehub:query"),
+		zcapld.WithCapabilityChain(o.cshURL+cshZCAP.ID), // TODO set cshZCAP.ID alone will cause panic in ZCAP
```

Comment:
Seems like the issue is in the CSH: https://github.com/trustbloc/edge-service/blob/574e2a43cb5a4d6cd1e7397c850ae32245beb7f9/pkg/restapi/csh/operation/operations.go#L144

Profile.ID should be a URN, not a string.

Reply:
done